enclave/src/lib.rs

```diff
@@ -522,136 +550,186 @@ where
-
-			// retrieve trusted operations from pool
-			let trusted_calls = author.get_pending_tops_separated(shard)?.0;
-
-			debug!("Got following trusted calls from pool: {:?}", trusted_calls);
-			// call execution
-			for trusted_call_signed in trusted_calls.into_iter() {
-				match handle_trusted_worker_call(
-					&mut calls,
-					&mut state,
-					trusted_call_signed,
-					latest_onchain_header.clone(),
-					shard,
-					Some(author.clone()),
-					on_chain_ocall,
-				) {
-					Ok(hashes) =>
-						if let Some((_, operation_hash)) = hashes {
-							call_hashes.push(operation_hash)
-						},
-					Err(e) => error!(
-						"Error performing worker call (will not push top hash): Error: {:?}",
-						e
-					),
-				};
-				// Check time
-				if time_is_overdue(Timeout::Call, start_time) {
-					is_done = true;
-					break
-				}
-			}
-			// create new block
-			match compose_block_and_confirmation(
+	let start_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() as i64;
+	let mut is_done = false;
+	for shard in shards.into_iter() {
+		let mut call_hashes = Vec::<H256>::new();
+
+		// load state before executing any calls
+		let mut state = load_initialized_state(&shard)?;
+		// save the state hash before call executions
+		// (needed for block composition)
+		let prev_state_hash = state::hash_of(state.state.clone())?;
+
+		// retrieve trusted operations from pool
+		let trusted_calls = author.get_pending_tops_separated(shard)?.0;
+
+		debug!("Got following trusted calls from pool: {:?}", trusted_calls);
+		// call execution
+		for trusted_call_signed in trusted_calls.into_iter() {
+			match handle_trusted_worker_call(
+				&mut calls,
+				&mut state,
+				&trusted_call_signed,
 				latest_onchain_header.clone(),
-				call_hashes,
 				shard,
-				prev_state_hash,
-				&mut state,
+				on_chain_ocall,
 			) {
-				Ok((block_confirm, signed_block)) => {
-					calls.push(block_confirm);
-					blocks.push(signed_block.clone());
-
-					// Notify watching clients of InSidechainBlock
-					let composed_block = signed_block.block();
-					let block_hash: BlockHash = blake2_256(&composed_block.encode()).into();
-					pool.pool()
-						.validated_pool()
-						.on_block_created(composed_block.signed_top_hashes(), block_hash);
+				Ok(hashes) => {
+					let inblock = match hashes {
+						Some((_, operation_hash)) => {
+							call_hashes.push(operation_hash);
+							true
+						},
+						None => {
+							// remove call as invalid from pool
+							false
+						},
+					};
+
+					// TODO: prune instead of remove_top ? Block needs to be known
+					// TODO: move this pruning to after finalization confirmations, not here!
+					// remove calls from pool (either as valid or invalid)
+					author
+						.remove_top(
+							vec![TrustedOperationOrHash::Operation(
+								trusted_call_signed.into_trusted_operation(true),
+							)],
+							shard,
+							inblock,
+						)
+						.unwrap();
 				},
```

Comment:
This part is extracted from the call into `handle_trusted_worker_call`. Instead of supplying an `Option<Author>`. Also, made the 3 instances where we remove calls from the TOP into 1, with the `inblock` variable that depends on the result of the `handle_trusted_worker_call` method.